### PR TITLE
feat(logicaldatabase): logical table expression parser

### PR DESCRIPTION
**Pull request by MarkPotato777**
#### What type of PR is this?

type-feature

#### What this PR does / why we need it:
This PR supports parsing logical table expressions.
1. The syntax for logic table expressions has been defined using ANTLR4, and the auto-generated `Visitor` class from ANTLR4 has been extended to implement specific business logic.
2. The syntax supports multiple logic table expressions connected by commas to avoid writing separate clause logic. At the business layer, the caller does not need to worry about whether it is a single expression or a list of expressions separated by commas; in fact, it can be considered as one expression either way.
3. More specifically, this leverages the `BaseStatement` abstraction defined by the `ob-sql-parser` to avoid some repetitive tasks, such as `getText`, `getStart`, `getStop`, etc.
4. The entry class is `LogicalTableService#resolve`, which is used to parse an expression into a` List<DataNode>`. Its implementation uses DefaultLogicalTableExpressionParser to parse and obtain `BaseLogicalTableExpression`, and then calls `BaseLogicalTableExpression#evaluate` to compute all the values derived from the expression.
5. The custom exception `BadExpressionException` has been defined for scenarios involving illegal expressions. This may be due to lexical or syntactic errors detected by ANTLR4 parsing, or could be semantic errors defined at the business layer, such as the requirement for the number of tables to be evenly divisible by the number of databases.
6. Furthermore, unit tests are divided into two parts: scenarios where the expression is invalid and scenarios where the expression is valid.

#### Which issue(s) this PR fixes:

Fixes #

#### Additional documentation e.g., usage docs, etc.:

```docs

```

### CODE REVIEW

File: server/odc-core/src/main/java/com/oceanbase/odc/core/shared/constant/ErrorCodes.java [hunk 1]
```diff
@@ -112,6 +112,11 @@ public enum ErrorCodes implements ErrorCode {
     ConnectHostNotAllowed,
     ConnectionTempOnly,
     ConnectionFlowConfigNotExists,
+    BadLogicalTableExpressionSyntax,
```
Comment (yizhouxw) [inline server/odc-core/src/main/java/com/oceanbase/odc/core/shared/constant/ErrorCodes.java:115]: all logic table related error code may start with LogicTable, as module name prefix

**Review by yizhouxw — COMMENTED** [review]
great job 👍

File: server/odc-core/src/main/java/com/oceanbase/odc/core/shared/constant/ErrorCodes.java (hunk 1 above)
Comment (MarkPotato777) [reply server/odc-core/src/main/java/com/oceanbase/odc/core/shared/constant/ErrorCodes.java]: okay~ Updated.

File: server/odc-service/src/main/java/com/oceanbase/odc/service/connection/logicaldatabase/BadExpressionException.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+import com.oceanbase.odc.core.shared.constant.ErrorCode;
+import com.oceanbase.odc.core.shared.exception.HttpException;
+
+/**
+ * @Author: Lebie
+ * @Date: 2024/4/23 14:03
+ * @Description: []
+ */
+public class BadExpressionException extends HttpException {
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/connection/logicaldatabase/BadExpressionException.java:28]: shall we move this exception to `com.oceanbase.odc.core.shared.exception`?

File: server/odc-service/src/main/java/com/oceanbase/odc/service/connection/logicaldatabase/parser/DefaultLogicalTableExpressionParser.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+import com.oceanbase.odc.service.connection.logicaldatabase.LogicalTableExpressionLexer;
+import com.oceanbase.odc.service.connection.logicaldatabase.LogicalTableExpressionParser;
+import com.oceanbase.tools.sqlparser.statement.Statement;
+
+/**
+ * @Author: Lebie
+ * @Date: 2024/4/23 10:13
+ * @Description: []
+ */
+public class DefaultLogicalTableExpressionParser {
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/connection/logicaldatabase/parser/DefaultLogicalTableExpressionParser.java:30]: why should we add the 'default' prefix?

may 'LogicalTableExpressionParser' instead~

File: server/odc-service/src/main/java/com/oceanbase/odc/service/connection/logicaldatabase/parser/DefaultLogicalTableExpressionParser.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    public Statement parse(String expression) throws SyntaxErrorException {
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/connection/logicaldatabase/parser/DefaultLogicalTableExpressionParser.java:31]: BaseSQLParser may help you and you can extend it~

File: server/odc-service/src/main/java/com/oceanbase/odc/service/connection/logicaldatabase/parser/SyntaxErrorException.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+import lombok.Getter;
+import lombok.NonNull;
+
+/**
+ * @Author: Lebie
+ * @Date: 2024/4/23 15:51
+ * @Description: []
+ */
+@Getter
+public class SyntaxErrorException extends ParseCancellationException {
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/connection/logicaldatabase/parser/SyntaxErrorException.java:37]: dup code

File: server/odc-service/src/main/java/com/oceanbase/odc/service/connection/logicaldatabase/parser/FastFailErrorListener.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+import org.antlr.v4.runtime.BaseErrorListener;
+import org.antlr.v4.runtime.RecognitionException;
+import org.antlr.v4.runtime.Recognizer;
+
+/**
+ * @Author: Lebie
+ * @Date: 2024/4/23 15:58
+ * @Description: []
+ */
+public class FastFailErrorListener extends BaseErrorListener {
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/connection/logicaldatabase/parser/FastFailErrorListener.java:27]: dup code

File: server/odc-service/src/main/java/com/oceanbase/odc/service/connection/logicaldatabase/parser/LogicalTableExpressionVisitor.java [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+    }
+
+    @Override
+    public BaseStatement visitTableSliceRangeWithBracket(TableSliceRangeWithBracketContext ctx) {
+        if (ctx.sliceRangeWithSingeBracket() != null) {
+            return visitSliceRangeWithSingeBracket(ctx.sliceRangeWithSingeBracket());
+        } else if (ctx.sliceRangeWithDoubleBracket() != null) {
+            return visitSliceRangeWithDoubleBracket(ctx.sliceRangeWithDoubleBracket());
+        } else {
+            return null;
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/connection/logicaldatabase/parser/LogicalTableExpressionVisitor.java:116]: this branch will never be run

File: server/odc-service/src/main/java/com/oceanbase/odc/service/connection/logicaldatabase/BadExpressionException.java (hunk 2 above)
Comment (MarkPotato777) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/connection/logicaldatabase/BadExpressionException.java:31]: I prefer to keep it in the service layer as it is a specific exception but not a general/common exception like `BadRepquest`, etc.

File: server/odc-service/src/main/java/com/oceanbase/odc/service/connection/logicaldatabase/parser/DefaultLogicalTableExpressionParser.java (hunk 3 above)
Comment (MarkPotato777) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/connection/logicaldatabase/parser/DefaultLogicalTableExpressionParser.java]: I added the 'default' prefix to distinguish `LogicalTableExpressionParser` generated by ANTLR4 itself.

File: server/odc-service/src/main/java/com/oceanbase/odc/service/connection/logicaldatabase/parser/DefaultLogicalTableExpressionParser.java (hunk 4 above)
Comment (MarkPotato777) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/connection/logicaldatabase/parser/DefaultLogicalTableExpressionParser.java]: done~

File: server/odc-service/src/main/java/com/oceanbase/odc/service/connection/logicaldatabase/parser/FastFailErrorListener.java (hunk 6 above)
Comment (MarkPotato777) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/connection/logicaldatabase/parser/FastFailErrorListener.java]: done~

File: server/odc-service/src/main/java/com/oceanbase/odc/service/connection/logicaldatabase/BadExpressionException.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+import com.oceanbase.odc.core.shared.constant.ErrorCodes;
+import com.oceanbase.odc.core.shared.exception.HttpException;
+import com.oceanbase.tools.sqlparser.SyntaxErrorException;
+
+/**
+ * @Author: Lebie
+ * @Date: 2024/4/23 14:03
+ * @Description: []
+ */
+public class BadExpressionException extends HttpException {
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/connection/logicaldatabase/BadExpressionException.java:31]: name suggestion: LogicalTableExpressionSyntaxErrorException

File: server/odc-service/src/main/java/com/oceanbase/odc/service/connection/logicaldatabase/parser/LogicalTableExpressionParseUtils.java [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+        List<String> result = new ArrayList<>();
+        for (int i = startInt; i <= endInt; i += stepInt) {
+            String format = includeLeadingZeros ? "%0" + (leadingZerosCount + String.valueOf(i).length()) + "d" : "%d";
+            result.add(String.format(format, i));
+        }
+
+        return result;
+    }
+
+    public static List<List<String>> cartesianProduct(List<List<String>> lists) {
```
Comment (yhilmare) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/connection/logicaldatabase/parser/LogicalTableExpressionParseUtils.java:74]: `com.oceanbase.odc.common.util.ListUtils#cartesianProduct`

File: server/odc-service/src/main/java/com/oceanbase/odc/service/connection/logicaldatabase/parser/LogicalTableExpressionVisitor.java [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+/**
+ * @Author: Lebie
+ * @Date: 2024/4/22 10:04
+ * @Description: []
+ */
+public class LogicalTableExpressionVisitor extends LogicalTableExpressionBaseVisitor<BaseStatement> {
+    @Override
+    public BaseStatement visitLogicalTableExpressionList(
+            LogicalTableExpressionParser.LogicalTableExpressionListContext ctx) {
+        PreConditions.notEmpty(ctx.logicalTableExpression(), "logicalTableExpression");
```
[the remark this answers is not in the document]
Comment (MarkPotato777) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/connection/logicaldatabase/parser/LogicalTableExpressionVisitor.java:50]: We'd better not trust the parameters passed from upstream blindly. Parameter validation is a defensive programming practice that does not result in any adverse effects.

File: server/odc-service/src/main/java/com/oceanbase/odc/service/connection/logicaldatabase/parser/LogicalTableExpressionVisitor.java [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+        expression.setSchemaExpression((SchemaExpression) visitSchemaExpression(ctx.schemaExpression()));
+        expression.setTableExpression((TableExpression) visitTableExpression(ctx.tableExpression()));
+        return expression;
+    }
+
+    @Override
+    public BaseStatement visitSchemaExpression(SchemaExpressionContext ctx) {
+        PreConditions.notEmpty(ctx.IDENTIFIER(), "schemaExpression.IDENTIFIER");
+        SchemaExpression expression = new SchemaExpression(ctx);
+        if (ctx.schemaSliceRangeWithBracket() != null) {
```
[the remark this answers is not in the document]
Comment (MarkPotato777) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/connection/logicaldatabase/parser/LogicalTableExpressionVisitor.java:74]: We'd better not trust the parameters passed from upstream blindly. Parameter validation is a defensive programming practice that does not result in any adverse effects.

File: server/odc-service/src/main/java/com/oceanbase/odc/service/connection/logicaldatabase/parser/LogicalTableExpressionVisitor.java [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+                ranges.add((BaseRangeExpression) visitTableSliceRangeWithBracket(sliceRange));
+            }
+            expression.setSliceRanges(ranges);
+        }
+        return expression;
+    }
+
+    @Override
+    public BaseStatement visitSchemaSliceRangeWithBracket(SchemaSliceRangeWithBracketContext ctx) {
+        if (ctx.sliceRangeWithSingeBracket() != null) {
```
[the remark this answers is not in the document]
Comment (MarkPotato777) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/connection/logicaldatabase/parser/LogicalTableExpressionVisitor.java:103]: We'd better not trust the parameters passed from upstream blindly. Parameter validation is a defensive programming practice that does not result in any adverse effects.

File: server/odc-service/src/main/java/com/oceanbase/odc/service/connection/logicaldatabase/parser/LogicalTableExpressionVisitor.java [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+        for (LogicalTableExpressionContext expression : ctx.logicalTableExpression()) {
+            expressionList.add((LogicalTableExpression) visitLogicalTableExpression(expression));
+        }
+        expressions.setExpressions(expressionList);
+        return expressions;
+    }
+
+    @Override
+    public BaseStatement visitLogicalTableExpression(LogicalTableExpressionContext ctx) {
+        PreConditions.notNull(ctx.schemaExpression(), "schemaExpression");
```
[the remark this answers is not in the document]
Comment (MarkPotato777) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/connection/logicaldatabase/parser/LogicalTableExpressionVisitor.java:62]: We'd better not trust the parameters passed from upstream blindly. Parameter validation is a defensive programming practice that does not result in any adverse effects.

File: server/odc-service/src/main/java/com/oceanbase/odc/service/connection/logicaldatabase/parser/LogicalTableExpressionVisitor.java [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+            range = (BaseRangeExpression) visitEnumRange(ctx.enumRange());
+        }
+        return range;
+    }
+
+    @Override
+    public BaseStatement visitConsecutiveRange(ConsecutiveRangeContext ctx) {
+        List<TerminalNode> numbers = ctx.NUMBER();
+        Verify.equals(2, numbers.size(), "consecutiveRange.NUMBER");
+        return new ConsecutiveSliceRange(ctx.getParent().getParent(), numbers.get(0).getText(),
```
[the remark this answers is not in the document]
Comment (MarkPotato777) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/connection/logicaldatabase/parser/LogicalTableExpressionVisitor.java:153]: Regarding a range, such as 1-100:2, what I actually want to obtain is [1-100:2], which makes subsequent replacement and displaying of error messages more convenient.

File: server/odc-service/src/main/java/com/oceanbase/odc/service/connection/logicaldatabase/parser/LogicalTableExpressionVisitor.java (hunk 7 above)
Comment (MarkPotato777) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/connection/logicaldatabase/parser/LogicalTableExpressionVisitor.java:116]: Anyway, this method need to return a value.

File: server/odc-service/src/main/java/com/oceanbase/odc/service/connection/logicaldatabase/parser/LogicalTableExpressionParseUtils.java (hunk 9 above)
Comment (MarkPotato777) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/connection/logicaldatabase/parser/LogicalTableExpressionParseUtils.java:74]: 👍🏻 updated.

File: server/odc-service/src/main/resources/antlr/logicaldatabase/LogicalTableExpression.g4 [hunk 15]
```diff
@@ -0,0 +1,39 @@
+grammar LogicalTableExpression;
```
[the remark this answers is not in the document]
Comment (MarkPotato777) [reply server/odc-service/src/main/resources/antlr/logicaldatabase/LogicalTableExpression.g4:17]: okay, done~

File: server/odc-service/src/main/java/com/oceanbase/odc/service/connection/logicaldatabase/parser/SyntaxErrorException.java (hunk 5 above)
Comment (MarkPotato777) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/connection/logicaldatabase/parser/SyntaxErrorException.java]: done

**Review by yhilmare — APPROVED** [review]
lgtm

**Review by JessieWuJiexi — APPROVED** [review]

**Review by smallsheeeep — APPROVED** [review]

**Review by PeachThinking — APPROVED** [review]
LGTM

**Pull request merged**